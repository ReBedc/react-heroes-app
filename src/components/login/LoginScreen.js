import { type } from '@testing-library/user-event/dist/type';
import React, { useContext, useReducer } from 'react'
import { useNavigate } from 'react-router-dom';
import { AuthContext } from '../../auth/authContext';
import { authReducer } from '../../auth/authReducer';
import { types } from '../../types/types';

export const LoginScreen = () => {

  const navigate = useNavigate(); // permite navegar a otras pantallas
  const { dispatch } = useContext(AuthContext);

  const handleLogin = () => {

    const action = {
      type: types.login,
      payload: {
        name: 'Fernando'
      }
    }
    dispatch(action);

    const lastPath = localStorage.getItem('lastPath') || '/marvel';

    navigate(lastPath, {
      replace: true // no regresa a la pantalla del login (reemplaza la ruta anterior en el stack)
    });
  }

  return (
    <div className="container mt-5">
        <h1>LoginScreen</h1>
        <hr />

        <button className="btn btn-primary"
          onClick={ handleLogin }>

          Login
        </button>
    </div>
  )
}