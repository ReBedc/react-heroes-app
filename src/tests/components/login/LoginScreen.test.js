import { mount } from "enzyme";
import { MemoryRouter } from "react-router-dom";
import { AuthContext } from "../../../auth/authContext";
import { LoginScreen } from "../../../components/login/LoginScreen";
import { types } from "../../../types/types";

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockNavigate
}));

describe('Pruebas con LoginScreen', () => {

    const contextValue = {
        dispatch: jest.fn(),
        user: {
            logged: false
        }
    }

    const wrapper = mount(
        <AuthContext.Provider value= {contextValue}>
            <MemoryRouter initialEntries={ ['/login'] }>
                <LoginScreen />
            </MemoryRouter>
        </AuthContext.Provider>
    );

    test('debe de mostrarse correctamente ', () => {

        const wrapper = mount(
            <AuthContext.Provider value= {contextValue}>
                <MemoryRouter initialEntries={ ['/login'] }>
                    <LoginScreen />
                </MemoryRouter>
            </AuthContext.Provider>
        );
        
        expect(wrapper).toMatchSnapshot();
        //expect(wrapper.find('.text-info').text().trim()).toBe('Juanito');
        //expect(wrapper.find('h1').text().trim()).toBe('MarvelScreen');

    });

    test('debe de realizar el dispatch y la navegacion', () => {

        const handleClick = wrapper.find('button').prop('onClick');
        handleClick();
        
        expect( contextValue.dispatch ).toHaveBeenCalledWith({
            type: types.login,
            payload: {name: 'Fernando'}
        });

        expect( mockNavigate ).toHaveBeenCalledWith('/marvel', {replace: true});

        localStorage.setItem('lastPath','/dc');
        handleClick();

        expect( mockNavigate ).toHaveBeenCalledWith('/dc', {replace: true});

        expect(wrapper).toMatchSnapshot();


    });

});