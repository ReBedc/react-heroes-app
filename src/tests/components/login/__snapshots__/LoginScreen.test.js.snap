// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas con LoginScreen debe de mostrarse correctamente  1`] = `
<div
  className="container mt-5"
>
  <h1>
    LoginScreen
  </h1>
  <hr />
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Login
  </button>
</div>
`;

exports[`Pruebas con LoginScreen debe de realizar el dispatch y la navegacion 1`] = `
<div
  className="container mt-5"
>
  <h1>
    LoginScreen
  </h1>
  <hr />
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Login
  </button>
</div>
`;
